flang/lib/Lower/OpenMP/OpenMP.cpp

```diff
@@ -1037,6 +1037,31 @@ static void genBodyOfTargetOp(
     genNestedEvaluations(converter, eval);
 }
 
+// If the symbol is specified in declare target directive, the function returns
+// the corresponding declare target operation.
+static mlir::omp::DeclareTargetInterface
+getDeclareTargetOp(const Fortran::semantics::Symbol &sym,
```

Comment:
Nice PR so far! :-)

However, I have a feeling this is only part of the puzzle with common blocks. I don't know enough about common blocks and OpenMPs interaction with them enough yet though, so I could most definitely be wrong. @mjklemm or @kiranchandramohan may perhaps know better in this case. 

However, if you emit the LLVM-IR for the two cases you've added to: openmp/libomptarget/test/offloading/fortran/declare-target-vars-in-target-region.f90 

You will see that the variables used in the regions are being treated as regular (implicitly) mapped variables, rather than declare target variables, and that the common block global is actually generated as a declare target global variable but unused on device. The elements of the common block are being accessed via a GEP and then passed to the target region, which seems reasonable from the perspective of infrastructure that's in place currently but I am not sure that's what we'd like in these cases.

I do not know if members of a declare target'd common block themselves should be treated as if they were declare target or not though (this looks from IR inspection to be something that could be quite difficult from the way common blocks are represented). From this specification exert on page 61 of the OpenMP 5.2 specification: 

"When a named common block appears in an OpenMP argument list, it has the same meaning and restrictions as if every explicit member of the common block appeared in the list. An explicit member of a common block is a variable that is named in a COMMON statement that specifies the common block name and is declared in the same scoping unit in which the clause appears. Named common blocks do not include the blank common block"

It seems like that may be the case, but my OpenMP-fu is trainee level at best, so please do verify for yourself and then get a second opinion! 

I think referring to a common block symbol directly in a map clause e.g. `map(tofrom: /CB/)` might be another good initial test if it is legal. From what I can tell it is, based on the definition of the map clause and locator lists (the latter also on page 61) but please do verify for yourself and if in doubt ask someone internally (or externally) who has more OpenMP specification knowledge than I do! The initial issue you will likely face with that test is that I do not believe it will go through the parser as I think it will complain about the symbol, I think it needs taught to deal with it like declare target currently does...

Reply:
Thanks @agozillon  for reviewing my PR. For the scalar, it seems to refer the common block global. However, for arrays it does not seem to. Let me dive a little deeper and will update this thread with my findings

Reply:
Thank you very much @anchuraj please do take the concerns I stated above with a grain of salt and feel free to tell me if I am wrong (as there's a good chance I am), it's thoughts from my previous work on declare target which didn't touch on common blocks other than to know it's something that OpenMP seems to require handling of. So at the very least it's something we can learn about together as we go :-)

Reply:
I think the main thing to keep in mind is that `variablename_ref_ptr` becomes the declare target variable rather than the original for the device IR and it's the thing that likely should be utilised to access the underlying data. This generated ref pointer is also utilised as the base pointer for the kernel entry info for the declare target argument as well in the host IR.

Reply:
Thanks @agozillon . References to common block structure in llvm ir is not correct. It is not appropriately looking at the right index even for scalar. As you said, we will have to update the code generation to refer to the right index in the common block global.  
 ```
1 PROGRAM TEST
  2
  3 INTEGER A
  4 INTEGER CBA
  5 COMMON /C1/ A, CBA
  6
  7 !$OMP DECLARE TARGET LINK(/C1/)
  8
  9 A=2
 10 CBA=100
 11 !$OMP TARGET
 12 A=A+1
 13 CBA=CBA+1
 14 !$OMP END TARGET
 15 PRINT *, CBA
 16 END
```
The above code prints wrong result in case of scalars as indexing is not done properly. 
I am cancelling this PR and will try to solve this problem .

Reply:
Thank you for looking into this @anchuraj, I think this might be a bit of a hard one to solve, but I've been wrong before (on a few occasions, and it's always a pleasant surprise). Do let me know if there is anything I can help with and I'll do my best, although, there is most definitely people that are more knowledgeable when it comes to common blocks.